2018/001-external-authentication.md

```diff
@@ -0,0 +1,125 @@
+
+This RFC proposes allowing Kubic operators to configure arbitary additional
+connectors within Kubic, pointing towards their own authentication backend.
+Initially, this will be limited to Dex's LDAP, SAML 2.0 and OpenID Connect
+connectors.
+
+This RFC does not propose removing that managed LDAP server, though that may
+be something we wish to pursue in the future.
+
+### Proposed change
+
+#### Velum Workflow Changes
+
+We propose to leave the initial bootstrap workflow unchanged, and instead
+allow additional authentication connectors to be configured post-bootstrap, in
+a similar manner to how we allow containter registries to be configured
+post-bootstap.
+
+Operators should be able to login to Velum, navigate to the Settings tab, and
+view, add or remove authentication connectors. Each authentication connector
+type (LDAP, SAML 2.0, OpenID Connect) will have a different view for the view
+and add pages, as Dex requires a different set of configuration inputs for
+each.
+
+It should be possible for operators to configure multiple connectors
+simultaneously, including multiple connectors of the same type (e.g. 2x LDAP).
+
+Once connectors have been configured, Velum will show an "Apply Changes" button,
+again - similar to container registries page, and Salt will be called to push
+out the updated configuration.
+
+#### Salt Implementation
+
+Velum will store the connector details in it's database, and will render the
+values into the Pillar API endpoint for Salt to read.
+
+Once in salt, most of the pieces are already in place for this. We'll need to
+update the Dex ConfigMap template (https://github.com/kubic-project/salt/blob/292b025681915a6ad4d7378624385d231f2f21d4/salt/addons/dex/manifests/15-configmap.yaml#L21)
+to render the new connector details, and then salt will apply the change to
+Kubernetes, causing a roll out of the change.
+
+#### CaaSP CLI Implementation
+
+As we now support SAML 2.0 and OpenID Connect, the current `caasp-cli` workflow
+will no longer be possible. CaaSP CLI must be updated to fully implement the
+OIDC browser based workflow. When a user logs in with `caasp-cli`, they must be
+presented with a URL to open in their browser, where they will login and copy
+the generated token back to a waiting `caasp-cli` in order to complete the
+authentication flow. The existing expectations that `caasp-cli` can provide the
+username and password on behalf of the user to Dex are no longer valid.
+
+### Dependencies
+
+This change should be reasonably isolated, simply providing the UI to enter the
+necessary data, and plumbing that through to Dex.
+
+However, documentation and testing efforts will be needed to ensure this feature
+is complete, and in working order.
+
+### Concerns and Unresolved Questions
+
+1. Dex will be restarted, in a rolling fashion, as the configmap is updated.
+   This may introduce authentication issues during the change.
+2. It's not known how Dex will handle an incorrectly configured connector,
```

Comment:
This is why we must include a test button on the auth config screen to validate the config. Given the risk we may break dex if incorrect details are added we must validate the config before updating dex IMHO